### ISSUE TITLE
Allow the ollama constrained decoding engine to fill in partially filled, and valid, json objects

### ISSUE THREAD

**Issue by willayy** [ollama/ollama#16674]
Allow the Ollama constrained decoding engine to fill in partially filled, and possible to make valid, JSON objects that are pre filled into the prompt. For example:
```py
import ollama

prefilled_prompt = """<|turn>system[Generate the next state given the current one. Your response must be a JSON object.]<|think|><turn|><|turn>user I am currently at Home. It is now 2026-06-10 08:15. What is my next state?<turn|><|turn>model<|channel>thought...<channel|>{"DateTime": "2026-06-10 08:15", """
# 2. The JSON Schema defining the remaining properties we want the LLM to choose
json_schema = {
    "type": "object",
    "properties": {
        "DateTime": {
            "type": "string",
        },
        "Activity": {
            "type": "string", 
            "enum": ["Sleeping", "Eating", "Idling"]
        }
    },
    "required": ["Activity","DateTime"],
    "additionalProperties": False
}
# 3. The request
response = ollama.generate(
    model="gemma4:e4b",
    prompt=prefilled_prompt,
    format=json_schema,
    raw=True
)
print(repr(response.response))
```
This would help me greatly in my research about state machines using LLMs since it allows me to save a lot of token generation time on values that could be generated quickly by some deterministic function. Also it should not be a hard thing to fix.

**Comment by rick-github** [comment]
The schema only applies to newly generated tokens.  Pre-filling with a partially completed structure is not handled because the runner doesn't know which token marks the start of the response.  A partial workaround is to tell the model the schema and then let it complete without the constraint:
```diff
--- 16674.py.orig	2026-06-11 12:09:04.336317235 +0200
+++ 16674.py	2026-06-11 12:09:49.866160729 +0200
@@ -1,6 +1,6 @@
 import ollama
 
-prefilled_prompt = """<|turn>system[Generate the next state given the current one. Your response must be a JSON object.]<|think|><turn|><|turn>user I am currently at Home. It is now 2026-06-10 08:15. What is my next state?<turn|><|turn>model<|channel>thought...<channel|>{"DateTime": "2026-06-10 08:15", """
+prefilled_prompt = """<|turn>system[Generate the next state given the current one. Your response must be a JSON object using this schema: {schema}.]<|think|><turn|><|turn>user I am currently at Home. It is now 2026-06-10 08:15. What is my next state?<turn|><|turn>model<|channel>thought...<channel|>{{"DateTime": "2026-06-10 08:15", """
 # 2. The JSON Schema defining the remaining properties we want the LLM to choose
 json_schema = {
     "type": "object",
@@ -19,8 +19,7 @@
 # 3. The request
 response = ollama.generate(
     model="gemma4:e4b",
-    prompt=prefilled_prompt,
-    format=json_schema,
+    prompt=prefilled_prompt.format(schema=json_schema),
     raw=True
 )
 print(repr(response.response))
```
```console
$ python3 16674.py
'\n"Activity": "Idling"}'
```
The issue is that without the constraint, the model is free to extend the response.

Getting this to work the way that you want would be an issue for [llama.cpp](https://github.com/ggml-org/llama.cpp/issues).

**Comment by willayy** [comment]
@rick-github Actually it does work... I must admit the reason I opened this issue was because the LLM chat bot im using told me this was the reason I was getting output that wasn't JSON. It said it wasn't possible... But it turns out it actually is very much possible, consider this example:

```py
import ollama

prefilled_prompt = """<|turn>system
Generate the next state given the current one. Your response must be a JSON object.<turn|>
<|turn>user
I am currently at Home.
It is now 2026-06-10 08:15.
What is my next state?<turn|>
<|turn>model
{"DateTime": "2026-06-10 08:15", """
# 2. The JSON Schema defining the remaining properties we want the LLM to choose
json_schema = {
    "type": "object",
    "properties": {
        "DateTime": {
            "type": "string",
        },
        "Activity": {
            "type": "string", 
            "enum": ["Sleeping", "Eating", "Idling"]
        }
    },
    "required": ["Activity","DateTime"],
    "additionalProperties": False
}
# 3. The request
response = ollama.generate(
    model="gemma4:e4b",
    prompt=prefilled_prompt,
    format=json_schema,
    raw=True
)
print(repr(response.response))
```

It works perfectly.

**Comment by rick-github** [comment]
It's not completing the partial response you gave it, it's generating a new one that matches the schema, using the data from the provided context.  But if that works for you, all good.

**Comment by willayy** [comment]
@rick-github im not sure im following. In my mind the LLM generates the next token using autoregression on top of the given prompt string and then when it generates the stop symbol Ollama just truncates from the last `<|turn>model ` symbol, excluding the ending `<turn|>` and returns that as  response.response. You're telling me this is not how it works? Would you mind filling me in😅

**Comment by rick-github** [comment]
By the time the generation has finished, the context buffer looks like this:
```<|turn>system
Generate the next state given the current one. Your response must be a JSON object.<turn|>
<|turn>user
I am currently at Home.
It is now 2026-06-10 08:15.
What is my next state?<turn|>
<|turn>model
{"DateTime": "2026-06-10 08:15", {"DateTime": "2026-06-10 08:15" , "Activity": "Idling" }
```
That is, the model is not completing your response, it's creating a new JSON structure compatible with the schema that contains the relevant data provided in the context.  If you wanted the model to complete the partial request provided in the assistant message, the output would be `"Activity": "Idling" }`, ie just the part of the JSON structure required to complete the partial supplied.  The only tokens sent to the clients are the ones generated, the runner does not work backwards to find the last `<|turn>model` and return every token after that point.

**Comment by willayy** [comment]
@rick-github  Ah right now i get it, thanks, would this be possible as a feature request? Or is too bad an architecture fit? This actually has a, albeit slim, use case.

**Comment by rick-github** [comment]
The architecture doesn't really support it.  By the time the generation starts, the input has been reduced to tokens and the runner has no idea what a token means, so is unable to find the point at which the partial assistant message starts.  Implementing this would required passing state from the frontend server to the backend runner, an operation not currently supported.  If it were to be implemented, it would be by the [llama.cpp](https://github.com/ggml-org/llama.cpp/issues) team as that is the backend runner doing the work.  When the MLX engine becomes mainstream, the work would have to be done in that runner as well.

It may be possible to fake it in the client.  The client knows where the partial messages starts, so it could do one token generation with logits, add a token that fits the schema from the returned logits, and then do another one token generation etc, until the response is complete.
